pkg/istio-agent/xds_proxy.go

```diff
@@ -708,3 +721,165 @@ func sendWithTimeout(ctx context.Context, sendFunc func(errorChan chan error)) e
 		return err
 	}
 }
+
+// tapRequest() sends "req" to Istiod, returning a channel through which the requestor gets their reply.
+//
+// Istiod won't reply with a unique ID to let us know when the
+// response has arrived, but will at least return the requested TypeUrl.  We will keep a map of lists of every
+// request for a TypeUrl, and when a response comes in we will alert the requesters first-come, first served.
+func (p *XdsProxy) tapRequest(req *discovery.DiscoveryRequest) (chan *discovery.DiscoveryResponse, error) {
+	if p.connected == nil {
+		return nil, fmt.Errorf("proxy not connected to Istiod")
+	}
+
+	req.ResponseNonce = time.Now().String()
+
+	responseChannel := make(chan *discovery.DiscoveryResponse)
+
+	p.tapMutex.Lock()
+	requestsQueue, ok := p.tapRequests[req.TypeUrl]
+	if !ok {
+		requestsQueue = make([]chan *discovery.DiscoveryResponse, 0, 1)
+	}
+	requestsQueue = append(requestsQueue, responseChannel)
+	p.tapRequests[req.TypeUrl] = requestsQueue
+
+	p.tapMutex.Unlock()
+
+	p.connected.requestsChan <- req
+
+	return responseChannel, nil
```

Comment:
if we do just have one enque-able, I think this function can also just block until it gets a response as well?